packages/graphback/src/layers/data/GraphbackDataProvider.ts

```diff
@@ -0,0 +1,35 @@
+
+// If we come with Union on client we might use some complex JSON for describing rules
+// and single key for type for simple use cases
+// tslint:disable-next-line: no-any
+export type AdvancedFilter = any;
+
+
+// TODO pagination
+interface PagedResult<T> {
+    result: T;
+    total: number;
+    page: number;
+}
+
+/**
+ * Builds CRUD interface for data access 
+ */
+// tslint:disable-next-line: no-any
+export interface GraphbackDataProvider<Type = any, GraphbackContext = any> {
+    // FIXME Pagination support
+    // CRUD
+    // FIXME rename to Object = object?
+    createObject(name: string, data: Type, context?: GraphbackContext): Promise<Type>;
+    readObject(name: string, id: string, context?: GraphbackContext): Promise<Type>;
+    updateObject(name: string, id: string, data: Type, context?: GraphbackContext): Promise<Type>;
+    deleteObject(name: string, id: string, context?: GraphbackContext): Promise<string>;
+
+    // Find
+    findAll(name: string, context?: GraphbackContext): Promise<Type[]>;
+    findBy(name: string, filter: Type | AdvancedFilter, context?: GraphbackContext): Promise<Type[]>;
+
+    // Dataloader support
+    // Caching support for nested queries
```

Comment:
@wtrocki I think the caching should be handled by the CRUDService, and for batching we should only expose one additional method `batchRead(name: string, ids: [string]): Promise<Type[]>`

Reply:
Very interesting. I like the batchRead idea in db level, however I was thinking about read with checking if id is already cached. Do you think we should be using some different approach?

As for data loading.. I think we can have composite (more than one service layers so loader will be there. I would want to add loader support only for nested data - where it makes sense.